### PR TITLE
feat: implement i18n for error messages #923 #924

### PR DESCRIPTION
**Pull request by ocruze**

### CODE REVIEW

File: src/SldStyleParser.ts [hunk 1]
```diff
@@ -65,6 +66,8 @@ export type ConstructorParams = {
   symbolizerUnits?: string;
   parserOptions?: X2jOptionsOptional;
   builderOptions?: XmlBuilderOptionsOptional;
+  translations?: Partial<SldStyleParserTranslations>;
```
Comment (KaiVolland) [inline src/SldStyleParser.ts:69]: If you use `Partial` you usually loose some informations and typings get less strict. Maybe you can just make the Locale keys optional too.

File: src/SldStyleParser.ts [hunk 2]
```diff
@@ -95,6 +98,52 @@ const COMBINATION_MAP = {
...
+  marksymbolizerParseFailedUnknownWellknownName?: string;
+  noFilterDetected?: string;
+  symbolizerKindParseFailed?: string;
+  colorMapEntriesParseFailedColorUndefined?: string;
+  contrastEnhancParseFailedHistoAndNormalizeMutuallyExclusive?: string;
+  channelSelectionParseFailedRGBAndGrayscaleMutuallyExclusive?: string;
+  channelSelectionParseFailedRGBChannelsUndefined?: string;
+};
+
+export type SldStyleParserTranslations = Record<Locale, Partial<SldStyleParserTranslationKeys>>;
```
Comment (KaiVolland) [inline src/SldStyleParser.ts:114]: As every key in the the `SldStyleParserTranslationKeys` is optional you should not need to use `Partial`.

File: src/SldStyleParser.ts [hunk 3]
```diff
@@ -210,9 +263,37 @@ export class SldStyleParser implements StyleParser<string> {
...
+        fr: {
+          translation: this.translations.fr,
+        }
+      }
+    });
+
     Object.assign(this, opts);
   }
 
+  translate(key: keyof SldStyleParserTranslationKeys, params?: any): string{
```
Comment (KaiVolland) [inline src/SldStyleParser.ts:293]:
```suggestion
  translate(key: keyof SldStyleParserTranslationKeys, params?: any): string {
```

File: src/SldStyleParser.ts [hunk 4]
```diff
@@ -95,6 +98,52 @@ const COMBINATION_MAP = {
 
 type CombinationType = keyof typeof COMBINATION_MAP;
 
+const SUPPORTED_LOCALES = ['en', 'de', 'fr'] as const;
+type Locale = (typeof SUPPORTED_LOCALES)[number];
```
Comment (KaiVolland) [inline src/SldStyleParser.ts:102]: I'm not sure if typing this is much of benefit as you can't provide a custom translation (e.g. spanish with 'es') without changing the source code. So I'd suggest to remove these typings:

```suggestion
```

File: src/SldStyleParser.ts (hunk 2 above)
Comment (KaiVolland) [inline src/SldStyleParser.ts:114]: See comment above regarding limiting translations.

```suggestion
export type SldStyleParserTranslations = Record<string, Partial<SldStyleParserTranslationKeys>>;
```

**Review by KaiVolland — COMMENTED** [review]
Nice job @ocruze. I added some notes to limitations the typings would introduce.

File: src/SldStyleParser.ts [hunk 5]
```diff
@@ -95,6 +98,49 @@ const COMBINATION_MAP = {
...
+      'Cannot parse ContrastEnhancement. Histogram and Normalize are mutually exclusive.',
+    channelSelectionParseFailedRGBAndGrayscaleMutuallyExclusive:
+      'Cannot parse ChannelSelection. RGB and Grayscale are mutually exclusive.',
+    channelSelectionParseFailedRGBChannelsUndefined: 
+      'Cannot parse ChannelSelection. Red, Green and Blue channels must be defined.'
+  },
+  de: {},
+  fr: {
+    marksymbolizerParseFailedUnknownWellknownName: 
+      'Lecture de MarkSymbolizer échoué. Le WellKnownName {{wellKnownName}} n\'est pas supporté.',
```
Comment (slafayIGN) [inline src/SldStyleParser.ts:130]: Échec de lecture du symbole de type MarkSymbolizer. Le WellKnownName...

File: src/SldStyleParser.ts [hunk 6]
```diff
@@ -95,6 +98,49 @@ const COMBINATION_MAP = {
...
+    noFilterDetected: 'Aucun filtre détecté.',
+    symbolizerKindParseFailed: 'Lecture de SymbolizerKind {{sldSymbolizerName}} à partir de SldRule échoué.',
```
Comment (slafayIGN) [inline src/SldStyleParser.ts:132]: Échec de lecture du type de symbole {{sldSymbolizerName}} à partir de SldRule.

File: src/SldStyleParser.ts [hunk 7]
```diff
@@ -95,6 +98,49 @@ const COMBINATION_MAP = {
...
+    colorMapEntriesParseFailedColorUndefined: 'Lecture de ColorMapEntries échoué. color n\'est pas défini.',
```
Comment (slafayIGN) [inline src/SldStyleParser.ts:133]: Échec...

File: src/SldStyleParser.ts [hunk 8]
```diff
@@ -95,6 +98,49 @@ const COMBINATION_MAP = {
...
+    contrastEnhancParseFailedHistoAndNormalizeMutuallyExclusive:
+      'Lecture de ContrastEnhancement échoué. Histogram et Normalize sont mutuellement exclusifs.',
```
Comment (slafayIGN) [inline src/SldStyleParser.ts:135]: Échec de lecture des propriétés de contraste ContrastEnhancement. Histogram....

File: src/SldStyleParser.ts [hunk 9]
```diff
@@ -95,6 +98,49 @@ const COMBINATION_MAP = {
...
+    channelSelectionParseFailedRGBAndGrayscaleMutuallyExclusive:
+      'Lecture de ChannelSelection échoué. RGB et Grayscale sont mutuellement exclusifs.',
```
Comment (slafayIGN) [inline src/SldStyleParser.ts:137]: Échec de lecture de la sélection de canaux ChannelSelection....

File: src/SldStyleParser.ts [hunk 10]
```diff
@@ -95,6 +98,49 @@ const COMBINATION_MAP = {
...
+    channelSelectionParseFailedRGBChannelsUndefined:
+      'Lecture de ChannelSelection échoué. Les canaux Rouge, Vert et Bleu doivent être définis.',
```
Comment (slafayIGN) [inline src/SldStyleParser.ts:139]: Pareil

**Review by slafayIGN — CHANGES_REQUESTED** [review]
Some other translations suggestions in french

**Review by jansule — APPROVED** [review]
Very nice!

**Pull request merged**